Src/Compilers/Core/MSBuildTask/BuildClient.cs

```diff
@@ -0,0 +1,398 @@
+        private const int TimeOutMsNewProcess = 20000; 
+
+        /// <summary>
+        /// Try to get the directory this assembly is in.
+        /// </summary>
+        private static string GetExpectedServerExeDir()
+        {
+            var uri = new Uri(Assembly.GetExecutingAssembly().CodeBase);
+            string assemblyPath = uri.IsFile 
+                ? uri.LocalPath
+                : Assembly.GetCallingAssembly().Location;
+            return Path.GetDirectoryName(assemblyPath);
+        }
+
+        /// <summary>
+        /// Returns a Task with a null BuildResponse if no server
+        /// response was received.
+        /// </summary>
+        public static Task<BuildResponse> TryRunServerCompilation(
+            RequestLanguage language,
+            string workingDir,
+            IList<string> arguments,
+            CancellationToken cancellationToken,
+            string libEnvVariable = null)
+        {
+            NamedPipeClientStream pipe;
+
+            var expectedServerExePath = Path.Combine(GetExpectedServerExeDir(), s_serverName);
+            var mutexName = expectedServerExePath.Replace('\\', '/');
+            bool mutexCreated;
+            using (var mutex = new Mutex(initiallyOwned: true,
+                                         name: mutexName,
+                                         createdNew: out mutexCreated))
+            {
+                var holdsMutex = mutexCreated;
+                if (!holdsMutex)
+                {
+                    try
+                    {
+                        holdsMutex = mutex.WaitOne(TimeOutMsNewProcess,
+                            exitContext: false);
+                    }
+                    catch (AbandonedMutexException)
+                    {
+                        holdsMutex = true;
+                    }
+                }
+
+                if (holdsMutex)
+                {
+                    var request = BuildRequest.Create(language, workingDir, Path.GetTempPath(), arguments, libEnvVariable);
+                    // Check for already running processes in case someone came in before us
+                    pipe = TryExistingProcesses(expectedServerExePath, cancellationToken);
+                    if (pipe != null)
+                    {
+                        return TryCompile(pipe, request, cancellationToken);
+                    }
+                    else
+                    {
+                        int processId = TryCreateServerProcess(expectedServerExePath);
+                        if (processId != 0 &&
+                            null != (pipe = TryConnectToProcess(processId,
+                                                                TimeOutMsNewProcess,
+                                                                cancellationToken)))
+                        {
+                            // Let everyone else access our process
+                            mutex.ReleaseMutex();
```

Comment:
Yeah, I mistakenly thought dispose releases the mutex. This should be in a finally block.